src/mono/mono/mini/method-to-ir.c

```diff
@@ -7563,17 +7565,34 @@ mono_method_to_ir (MonoCompile *cfg, MonoMethod *method, MonoBasicBlock *start_b
 								g_array_append_val (new_params, lowered_swift_struct.lowered_elements [idx_lowered]);
 							}
 						} else {
-							// For structs that cannot be lowered, we change the argument to byref type
-							ptype = mono_class_get_byref_type (mono_defaults.typed_reference_class);
-							// Load the address of the struct
-							MonoInst *struct_base_address =  mono_compile_create_var (cfg, mono_get_int_type (), OP_LOCAL);
-							CHECK_ARG (idx_param);
-							NEW_ARGLOADA (cfg, struct_base_address, idx_param);
-							MONO_ADD_INS (cfg->cbb, struct_base_address);
-							*sp++ = struct_base_address;
+							if (gklass && (gklass->container_class == swift_self_t)) {
+								// For SwiftSelf<T> that cannot be lowered, we change the argument to byref SwiftSelf type
+								ptype = mono_class_get_byref_type (swift_self);
+								MonoInst *swift_self_base_address = mono_compile_create_var (cfg, mono_get_int_type (), OP_LOCAL);
+								CHECK_ARG (idx_param);
+								NEW_ARGLOADA (cfg, swift_self_base_address, idx_param);
+								MONO_ADD_INS (cfg->cbb, swift_self_base_address);
+
+								MonoClassField *klass_fields = m_class_get_fields (klass);
+								MonoInst *struct_base_address = mono_compile_create_var (cfg, mono_get_int_type (), OP_LOCAL);
+								MONO_EMIT_NEW_LOAD_MEMBASE_OP (cfg, OP_ADD_IMM, struct_base_address->dreg, swift_self_base_address->dreg, klass_fields->offset);
```

Comment:
NIT: Maybe just this part could be in an `if` block to avoid code duplication?

Reply:
Let me know if it is still readable and understandable. If there are too many branches, we can revert it.

Reply:
This look cleaner for me now